File: analysis/BrushDynamics.java
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Brush stroke dynamics estimator aligned with the Lua reference implementation.
 */
public final class BrushDynamics {
    private static final double WIDTH = 0.40;      // m
    private static final double HEIGHT = 0.79;     // m
    private static final double TEXTURE_MM = 20.0; // mm cap
    private static final double BRUSH_MASS = 0.12; // kg
    private static final double GRAVITY = 9.81;    // m/s^2
    private static final double MU_BASE = 0.32;
    private static final double MU_TEXTURE_GAIN = 0.06;
    private static final double LONG_TIME = 0.45;  // s
    private static final double SHORT_TIME = 0.35; // s
    private static final double ROTATION_ANGLE = Math.PI; // rad (180 deg)

    private BrushDynamics() {}

    private static double effectiveFriction() {
        double normalized = Math.min(TEXTURE_MM, 20.0) / 20.0;
        return MU_BASE + MU_TEXTURE_GAIN * normalized;
    }

    private static double symmetricAccel(double distance, double duration) {
        return 4.0 * distance / (duration * duration);
    }

    private static Map<String, Double> linearMetrics(double length, double duration, double mass,
                                                     double mu, double g) {
        double speed = length / duration;
        double acceleration = symmetricAccel(length, duration);
        double tension = mass * (acceleration + mu * g);
        Map<String, Double> metrics = new LinkedHashMap<>();
        metrics.put("length", length);
        metrics.put("duration", duration);
        metrics.put("speed", speed);
        metrics.put("acceleration", acceleration);
        metrics.put("tension", tension);
        return metrics;
    }

    private static Map<String, Double> rotationMetrics(double radius, double theta, double duration,
                                                       double mass, double mu, double g) {
        double arc = radius * theta;
        double speed = arc / duration;
        double angularSpeed = speed / radius;
        double tangentialAccel = symmetricAccel(arc, duration);
        double centripetalAccel = speed * speed / radius;
        double tension = mass * (tangentialAccel + centripetalAccel + mu * g);
        Map<String, Double> metrics = new LinkedHashMap<>();
        metrics.put("radius", radius);
        metrics.put("arc", arc);
        metrics.put("duration", duration);
        metrics.put("speed", speed);
        metrics.put("angular_speed", angularSpeed);
        metrics.put("tangential_accel", tangentialAccel);
        metrics.put("centripetal_accel", centripetalAccel);
        metrics.put("tension", tension);
        return metrics;
    }

    private static void printMetrics(String label, Map<String, Double> metrics) {
        System.out.println(label + ":");
        metrics.forEach((key, value) -> System.out.printf("  %-18s %.4f%n", key, value));
    }

    public static void main(String[] args) {
        double mu = effectiveFriction();
        System.out.printf("Effective friction coefficient: %.4f%n", mu);

        Map<String, Double> linearLong = linearMetrics(HEIGHT, LONG_TIME, BRUSH_MASS, mu, GRAVITY);
        Map<String, Double> linearShort = linearMetrics(WIDTH, SHORT_TIME, BRUSH_MASS, mu, GRAVITY);

        double halfDiagonal = Math.hypot(WIDTH * 0.5, HEIGHT * 0.5);
        Map<String, Double> rotationLong = rotationMetrics(halfDiagonal, ROTATION_ANGLE, LONG_TIME,
                BRUSH_MASS, mu, GRAVITY);
        Map<String, Double> rotationShort = rotationMetrics(WIDTH * 0.5, ROTATION_ANGLE, SHORT_TIME,
                BRUSH_MASS, mu, GRAVITY);

        printMetrics("Linear long stroke", linearLong);
        printMetrics("Linear short stroke", linearShort);
        printMetrics("Rotational long sweep", rotationLong);
        printMetrics("Rotational short sweep", rotationShort);
    }
}

File: analysis/brush_dynamics.lua
-- Brush stroke dynamics estimator for a 40cm x 79cm linen canvas
-- All units SI unless stated otherwise.

local canvas = {
    width = 0.40,   -- meters
    height = 0.79   -- meters
}

local params = {
    texture_mm = 20.0,
    brush_mass = 0.12, -- kg
    gravity = 9.81,
    mu_base = 0.32,
    mu_texture_gain = 0.06,
    long_time = 0.45,  -- seconds
    short_time = 0.35, -- seconds
    rotation_angle = math.pi -- radians (180 deg sweep)
}

local function effective_friction(mu_base, gain, texture_mm)
    local normalized = math.min(texture_mm, 20.0) / 20.0
    return mu_base + gain * normalized
end

local function symmetric_accel(distance, duration)
    return 4.0 * distance / (duration * duration)
end

local function linear_metrics(length, duration, mass, mu, g)
    local speed = length / duration
    local accel = symmetric_accel(length, duration)
    local tension = mass * (accel + mu * g)
    return {
        length = length,
        duration = duration,
        speed = speed,
        acceleration = accel,
        tension = tension
    }
end

local function rotation_metrics(radius, theta, duration, mass, mu, g)
    local arc = radius * theta
    local speed = arc / duration
    local tangential_accel = symmetric_accel(arc, duration)
    local centripetal_accel = speed * speed / radius
    local tension = mass * (tangential_accel + centripetal_accel + mu * g)
    return {
        radius = radius,
        arc = arc,
        duration = duration,
        speed = speed,
        angular_speed = speed / radius,
        tangential_accel = tangential_accel,
        centripetal_accel = centripetal_accel,
        tension = tension
    }
end

local mu = effective_friction(params.mu_base, params.mu_texture_gain, params.texture_mm)

local linear_long = linear_metrics(canvas.height, params.long_time, params.brush_mass, mu, params.gravity)
local linear_short = linear_metrics(canvas.width, params.short_time, params.brush_mass, mu, params.gravity)

local half_diagonal = math.sqrt((canvas.width * 0.5)^2 + (canvas.height * 0.5)^2)
local rotation_long = rotation_metrics(half_diagonal, params.rotation_angle, params.long_time, params.brush_mass, mu, params.gravity)
local rotation_short = rotation_metrics(canvas.width * 0.5, params.rotation_angle, params.short_time, params.brush_mass, mu, params.gravity)

local function format_metrics(label, keys, data)
    print(string.format("%s:", label))
    for _, key in ipairs(keys) do
        local value = data[key]
        if value then
            print(string.format("  %-18s %.4f", key, value))
        end
    end
end

local linear_keys = { "length", "duration", "speed", "acceleration", "tension" }
local rotation_keys = {
    "radius",
    "arc",
    "duration",
    "speed",
    "angular_speed",
    "tangential_accel",
    "centripetal_accel",
    "tension"
}

print(string.format("Effective friction coefficient: %.4f", mu))
format_metrics("Linear long stroke", linear_keys, linear_long)
format_metrics("Linear short stroke", linear_keys, linear_short)
format_metrics("Rotational long sweep", rotation_keys, rotation_long)
format_metrics("Rotational short sweep", rotation_keys, rotation_short)
